packages/app/src/components/PatternTable.tsx

```diff
@@ -0,0 +1,258 @@
+import { useMemo } from 'react';
+import stripAnsi from 'strip-ansi';
+import { ChartConfigWithDateRange } from '@hyperdx/common-utils/dist/types';
+import { useQuery } from '@tanstack/react-query';
+
+import {
+  convertDateRangeToGranularityString,
+  timeBucketByGranularity,
+  toStartOfInterval,
+} from '@/ChartUtils';
+import { useQueriedChartConfig } from '@/hooks/useChartConfig';
+
+import {
+  RawLogTable,
+  selectColumnMapWithoutAdditionalKeys,
+  useConfigWithPrimaryAndPartitionKey,
+} from './DBRowTable';
+import { useSearchTotalCount } from './SearchTotalCountChart';
+
+// We don't want to load pyodide over and over again, use react query to cache the async instance
+function usePyodide() {
+  return useQuery({
+    queryKey: ['pyodide'],
+    queryFn: async () => {
+      // @ts-ignore
+      const pyodide = await window.loadPyodide();
+      await pyodide.loadPackage('micropip');
+      const micropip = pyodide.pyimport('micropip');
+      const url = new URL(
+        '/drain3-0.9.11-py3-none-any.whl',
+        window.location.origin,
+      );
+      await micropip.install(url.href);
+      return pyodide;
+    },
+    refetchOnWindowFocus: false,
+    refetchOnMount: false,
+    refetchInterval: false,
+  });
+}
+
+async function mineEventPatterns(logs: string[], pyodide: any) {
+  pyodide.globals.set('HDXLOGS', logs);
+  return JSON.parse(
+    await pyodide.runPythonAsync(`
+import js
+import json
+from drain3 import TemplateMiner
+from drain3.template_miner_config import TemplateMinerConfig
+m = TemplateMiner(None, TemplateMinerConfig())
+results = []
+for log in HDXLOGS:
+    results.append(m.add_log_message(log))
+
+json.dumps(results)
+    `),
+  );
+}
+
```

Comment:
Two things, neither are critical right now:
1) It would be nice if we put this code in an actual `.py` file and import/consume it here. Just easier to maintain -- although I don't anticipate many changes, hence it feels like an overoptimization right now.

2) Second...maybe we wrap this in a try/catch so we can more gracefully show an error message?

```
try {
  await pyodide.runPythonAsync(code);
} catch (e) {
  console.error('Pyodide error:', e);
  setErrorState('Failed to parse patterns');
}
```

Reply:
1. yeah I think this is tricky as I don't think the Next bundler will like py files, we could ofc add a constants file that contains python code but I think that makes this less readable.
2. Building on below I think we should throw here and let the caller (react-query) deal with errors instead.